Validation/SiTrackerPhase2V/plugins/Phase2ITValidateCluster.cc

```diff
@@ -0,0 +1,372 @@
+            if (it != clusterSimTrackIds.end() && *it == simhitIt.trackId()) {
+              if (!closestSimHit || fabs(simhitIt.localPosition().x() - localPosCluster.x()) < minx) {
+                minx = abs(simhitIt.localPosition().x() - localPosCluster.x());
+                closestSimHit = &simhitIt;
+              }
+            }
+          }
+        }  //end loop over PSimhitcontainers
+      }    //end loop over simHits
+
+      if (!closestSimHit)
+        continue;
+      // only look at simhits from highpT tracks
+      auto simTrackIt(simTracks.find(closestSimHit->trackId()));
+      if (simTrackIt == simTracks.end())
+        continue;
+      Local3DPoint localPosSimHit(closestSimHit->localPosition());
+      const double deltaX = localPosCluster.x() - localPosSimHit.x();
+      const double deltaY = localPosCluster.y() - localPosSimHit.y();
+
+      auto layerMEIt = layerMEs_.find(folderkey);
+      if (layerMEIt == layerMEs_.end())
+        continue;
+
+      ClusterMEs& local_mes = layerMEIt->second;
+      local_mes.deltaX_P->Fill(deltaX);
+      local_mes.deltaY_P->Fill(deltaY);
+      // Primary particles only
+      if (phase2tkutil::isPrimary(simTrackIt->second, closestSimHit)) {
+        local_mes.deltaX_P_primary->Fill(deltaX);
+        local_mes.deltaY_P_primary->Fill(deltaY);
+      }
+    }
+  }
+}
+
+//
+// -- Book Histograms
+//
+void Phase2ITValidateCluster::bookHistograms(DQMStore::IBooker& ibooker,
+                                             edm::Run const& iRun,
+                                             edm::EventSetup const& iSetup) {
+  std::string top_folder = config_.getParameter<std::string>("TopFolderName");
+  edm::LogInfo("Phase2ITValidateCluster") << " Booking Histograms in: " << top_folder;
+
+  edm::ESWatcher<TrackerDigiGeometryRecord> theTkDigiGeomWatcher;
+  if (theTkDigiGeomWatcher.check(iSetup)) {
+    for (auto const& det_u : tkGeom_->detUnits()) {
+      //Always check TrackerNumberingBuilder before changing this part
+      if (!(det_u->subDetector() == GeomDetEnumerators::SubDetector::P2PXB ||
+            det_u->subDetector() == GeomDetEnumerators::SubDetector::P2PXEC))
+        continue;  //continue if not Pixel
+      uint32_t detId_raw = det_u->geographicalId().rawId();
+      bookLayerHistos(ibooker, detId_raw, top_folder);
+    }
+  }
+}
+
+//////////////////Layer Histo/////////////////////////////////
+void Phase2ITValidateCluster::bookLayerHistos(DQMStore::IBooker& ibooker, uint32_t det_id, const std::string& subdir) {
+  std::string folderName = phase2tkutil::getITHistoId(det_id, tTopo_);
+  if (folderName.empty()) {
+    edm::LogWarning("Phase2ITValidateCluster") << ">>>> Invalid histo_id ";
+    return;
+  }
+
+  if (layerMEs_.find(folderName) == layerMEs_.end()) {
+    ibooker.cd();
+    ibooker.setCurrentFolder(subdir + '/' + folderName);
+    edm::LogWarning("Phase2TrackerValidateDigi") << " Booking Histograms in: " << subdir + '/' + folderName;
```

Comment:
@sroychow as you need anyway to touch the code again to reduce the memory, please make this `edm::LogWarning` into a `LogInfo`. This doesn't need to be a warning and it shows in DQM bin by bin comparison: https://tinyurl.com/y5mnseq2

Reply:
Done